### PR TITLE
packages/base-files: Eliminate firstboot overlay race

### PR DESCRIPTION
**Pull request by daniel-santos**
Hello!  It's been a little while since I've submitted an OpenWRT patch.  This patch increases time required for firstboot by a non-trivial amount but closes a race condition where data potentially can fall into a bit bucket.

Using tempfs as an overlay upon first boot is a nice way to speed it up,
but it is inherently racy. This is because we don't have control over
what processes will be started in the processing of /etc/rc.d/S* and
these processes could have open files in the tempfs overlay.

Thus, when /etc/rc.d/S95done is run, and runs `mount_root done` where
jffs2_switch() [1] is called, it copies the files in /tmp/root/upper,
switches the filesystem and then copies the files again to "try hard to
be in sync". While this certainly increases the changes of being in
sync, it cannot guarantee it because files can still be open and
continue to be modified. A process may have writes pending in it's
process space that are buffered and not yet flushed to the file system,
etc.

This patch is an intermediate solution that eliminates the race and
provides integrity guarantee.

[1] https://git.openwrt.org/?p=project/fstools.git;a=blob;f=libfstools/overlay.c;h=67903372d567602cac20fbab0fdc64cbabc1acde;hb=93369be040612c906bcbb1631f44a92fa4122d24#l316

Signed-off-by: Daniel Santos <daniel@gsat.us>

### CODE REVIEW

**Comment by daniel-santos** [comment]
I should mention that patch does not remove the code is obsoletes, which should be done for housekeeping.

Also, I'm sure there's better solutions to this problem, but it's at least the simplest.  I don't know jffs2 internals, but instead of taring up a `sysupgrade.tgz`, it might be better to just create a new jffs2 file system during `sysupgrade` (without `-n`).  But this is better discussed in the mailing list.

**Comment by ptpt52** [comment]
mark

**Comment by ptpt52** [comment]
I'm confused
why it `rm -f /sysupgrade.tgz` in `80_mount_root` and also do `rm -f /sysupgrade.tgz` in /etc/init.d/done?

**Comment by daniel-santos** [comment]
> I'm confused why it `rm -f /sysupgrade.tgz` in `80_mount_root` and also do `rm -f /sysupgrade.tgz` in /etc/init.d/done?

Hello.  I'm not seeing there ` /sysupgrade.tgz` is deleted in `/lib/preinit/80_mount_root`.  When `mount_root` is called without any arguments I believe it moves it from the jffs2 file system that sysupgrade creates and into the the tmpfs overlay, but when `mount_root done` is called, it moves it from tmpfs back to the new jffs2 overlay.  Honestly, it's a little hard to follow, but there's not an `rm -f /sysupgrade.tgz` in the shell script its self.

**Comment by ptpt52** [comment]
would this change affected on initramfs-kernel boot up?
